File: app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkNavigator.java
package net.squanchy.navigation.deeplink;

import java.util.ArrayList;
import java.util.List;

import net.squanchy.navigation.Navigator;
import net.squanchy.support.lang.Optional;

class DeepLinkNavigator {

    private final Navigator navigator;

    DeepLinkNavigator(Navigator navigator) {
        this.navigator = navigator;
    }

    void toSchedule(String path) {
        List<String> pathSegments = extractSegments(path);

        Optional<String> dayId = Optional.absent();
        if (!pathSegments.isEmpty()) {
            dayId = Optional.of(pathSegments.get(0));
        }

        Optional<String> eventId = Optional.absent();
        if (pathSegments.size() > 1) {
            eventId = Optional.of(pathSegments.get(1));
        }
        navigator.toSchedule(dayId, eventId);
    }

    void toEventDetails(String path) {
        List<String> pathSegments = extractSegments(path);
        if (pathSegments.isEmpty()) {
            throw DeepLinkParsingException.eventDetailsMissingEventId(path);
        }

        String eventId = pathSegments.get(0);
        navigator.toEventDetails(eventId);
    }

    void toSpeakerUri(String path) {
        List<String> pathSegments = extractSegments(path);
        if (pathSegments.isEmpty()) {
            throw DeepLinkParsingException.speakerDetailsMissingSpeakerId(path);
        }

        String speakerId = pathSegments.get(0);
        navigator.toSpeakerDetails(speakerId);
    }

    private List<String> extractSegments(String path) {
        String[] allSegments = path.split("/");
        ArrayList<String> filteredSegments = new ArrayList<>();
        for (String segment : allSegments) {
            if (!segment.isEmpty()) {
                filteredSegments.add(segment);
            }
        }
        return filteredSegments;
    }

    void toFavorites() {
        navigator.toFavorites();
    }

    void toSocial() {
        navigator.toTwitterFeed();
    }

    void toVenue() {
        navigator.toVenueInfo();
    }

    void toDebug() {
        navigator.toDebugSettings();
    }
}

File: app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkNavigator.kt
package net.squanchy.navigation.deeplink

import net.squanchy.navigation.Navigator

internal class DeepLinkNavigator(private val navigator: Navigator) {

    fun toSchedule(path: String) {
        val pathSegments = extractSegments(path)

        val dayId = if (pathSegments.isNotEmpty()) pathSegments.first() else null
        val eventId = if (pathSegments.size > 1) pathSegments[1] else null

        navigator.toSchedule(dayId, eventId)
    }

    fun toEventDetails(path: String) {
        val pathSegments = extractSegments(path)
        if (pathSegments.isEmpty()) {
            throw DeepLinkParsingException.eventDetailsMissingEventId(path)
        }

        val eventId = pathSegments.first()
        navigator.toEventDetails(eventId)
    }

    fun toSpeakerUri(path: String) {
        val pathSegments = extractSegments(path)
        if (pathSegments.isEmpty()) {
            throw DeepLinkParsingException.speakerDetailsMissingSpeakerId(path)
        }

        val speakerId = pathSegments.first()
        navigator.toSpeakerDetails(speakerId)
    }

    private fun extractSegments(path: String): List<String> {
        return path.split("/".toRegex())
            .filterNot { it.isEmpty() }
    }

    fun toFavorites() {
        navigator.toFavorites()
    }

    fun toSocial() {
        navigator.toTwitterFeed()
    }

    fun toVenue() {
        navigator.toVenueInfo()
    }

    fun toDebug() {
        navigator.toDebugSettings()
    }
}
